src/garage/torch/algos/sac.py

```diff
@@ -0,0 +1,258 @@
+                         use_target=True,
+                         discount=discount,
+                         smooth_return=smooth_return)
+
+        # use 2 target q networks
+        self.target_qf1 = copy.deepcopy(self.qf1)
+        self.target_qf2 = copy.deepcopy(self.qf2)
+        # self.target_qf1 = target_qf1
+        # self.target_qf2 = target_qf2
+        self.policy_optimizer = optimizer(self.policy.parameters(),
+                                          lr=self.policy_lr)
+        self.qf1_optimizer = optimizer(self.qf1.parameters(), lr=self.qf_lr)
+        self.qf2_optimizer = optimizer(self.qf2.parameters(), lr=self.qf_lr)
+        # automatic entropy coefficient tuning
+        self.use_automatic_entropy_tuning = use_automatic_entropy_tuning
+        if self.use_automatic_entropy_tuning and not alpha:
+            if target_entropy:
+                self.target_entropy = target_entropy
+            else:
+                self.target_entropy = -np.prod(
+                    self.env_spec.action_space.shape).item()
+            self.log_alpha = torch.tensor([self.initial_log_entropy], dtype=torch.float, requires_grad=True)
+            self.alpha_optimizer = optimizer([self.log_alpha], lr=self.policy_lr)
+        else:
+            self.alpha = alpha
+
+        self.episode_rewards = []
+        self.success_history = []
+
+    # 0) update policy using updated min q function
+    # 1) compute targets from Q functions
+    # 2) update Q functions using optimizer
+    # 3) query Q functons, take min of those functions
+    def train_once(self, itr, paths):
+        """
+        """
+        paths = self.process_samples(itr, paths)
+        self.episode_rewards.extend([
+            path for path, complete in zip(paths['undiscounted_returns'],
+                                           paths['complete']) if complete
+        ])
+        self.success_history.extend([
+            path for path, complete in zip(paths['success_history'],
+                                           paths['complete']) if complete
+        ])
+        last_average_return = np.mean(self.episode_rewards)
+        if self.replay_buffer.n_transitions_stored >= self.min_buffer_size:  # noqa: E501
+            for gradient_step in range(self.gradient_steps):
+                    samples = self.replay_buffer.sample(self.buffer_batch_size)
+                    # self.update_q_functions(itr, samples)
+                    self.optimize_policy(gradient_step, samples)
+                    self.update_targets()
+            tabular.record('reward', last_average_return)
+
+        return last_average_return
+
+    def optimize_policy(self, itr, samples):
+        """ Optimize the policy based on the policy objective from the sac paper.
+
+        Args:
+            itr (int) - current training iteration
+            samples() - samples recovered from the replay buffer
+        Returns:
+            None
+        """
+
```

Comment:
You can use np_to_torch from garage.torch.utils on samples here to convert them to torch tensors, so you have have to keep wrapping everything in torch.Tensors later on.